freqtrade/freqai/data_drawer.py

```diff
@@ -78,25 +79,53 @@ def __init__(self, full_path: Path, config: Config, follow_mode: bool = False):
         self.historic_predictions_bkp_path = Path(
             self.full_path / "historic_predictions.backup.pkl")
         self.pair_dictionary_path = Path(self.full_path / "pair_dictionary.json")
+        self.metric_tracker_path = Path(self.full_path / "metric_tracker.json")
         self.follow_mode = follow_mode
         if follow_mode:
             self.create_follower_dict()
         self.load_drawer_from_disk()
         self.load_historic_predictions_from_disk()
+        self.load_metric_tracker_from_disk()
         self.training_queue: Dict[str, int] = {}
         self.history_lock = threading.Lock()
         self.save_lock = threading.Lock()
         self.pair_dict_lock = threading.Lock()
+        self.metric_tracker_lock = threading.Lock()
         self.old_DBSCAN_eps: Dict[str, float] = {}
         self.empty_pair_dict: pair_info = {
                 "model_filename": "", "trained_timestamp": 0,
                 "data_path": "", "extras": {}}
+        self.metric_tracker: Dict[str, Dict[str, list]] = {}
+
+    def update_metric_tracker(self, metric: str, value: float, pair: str) -> None:
+        """
+        General utility for adding and updating custom metrics. Typically used
+        for adding training performance, train timings, inferenc timings, cpu loads etc.
+        """
+        with self.metric_tracker_lock:
+            if pair not in self.metric_tracker:
+                self.metric_tracker[pair] = {}
+            if metric not in self.metric_tracker[pair]:
+                self.metric_tracker[pair][metric] = []
+
+            self.metric_tracker[pair][metric].append(value)
```

Comment:
Opposing my other comment on data size - i don't think what you're currently saving makes sense.

it'll end as an array of "train_time, cpu1, cpu5, cpu15" - per pair - but it's lacking any time-reference (you'll have a datapoint, but you'll only know it's the 15th datapoint (measured from when?), not "this happened at 11:45 on the 15th october".

This will make correlating this to anything quite difficult / impossible - which makes me doubt why we'd need to save this info in the first place.

Reply:
We don't care about *when* these are called. These are for understanding statistical representations of these metrics.

Inference timings vector wont be the same length as train timings vector.